Mage/src/main/java/mage/game/permanent/token/TokenImpl.java

```diff
@@ -316,6 +313,11 @@ private static void putOntoBattlefieldHelper(CreateTokenEvent event, Game game,
                 // tokens zcc must simulate card's zcc to keep copied card/spell settings
                 // (example: etb's kicker ability of copied creature spell, see tests with Deathforge Shaman)
                 newPermanent.updateZoneChangeCounter(game, emptyEvent);
+
+                if (source != null && source.getCostsTagMap() != null) {
+                    MageObjectReference mor = new MageObjectReference(newPermanent.getId(),newPermanent.getZoneChangeCounter(game)-1,game);
+                    game.getPermanentCostsTags().put(mor, source.getCostsTagMap());
```

Comment:
1. Must have `initPermanentCostsTags(name, sourceCostsTags)` instead `put` usage. And use deep copy inside and add usage comments (e.g. must init tags on spell/cast resolve only).

2. There are "miss" usages like bestow to battlefield -- is it affected or not?
![shot_231113_151430](https://github.com/magefree/mage/assets/8344157/637c654f-b8a4-4f2f-b1c3-43770981c93d)
![shot_231113_151443](https://github.com/magefree/mage/assets/8344157/0ec6a9c6-00dc-4444-841a-e214ccb0e494)

Reply:
1. `game.storePermanentCostsTags(MageObjectReference permanentMOR, Ability source)` sound good? I'll add more comments. As for deep copy, personally I prefer requiring immutable values but always deep copying is more consistent with existing code so I'll switch to that and remove the comments mentioning requiring immutables.
2. I don't think there are any cards where it could matter but you're right, Bestow without target should store the tags as well. The second one should not since resolved instant/sorcery cards going to the graveyard aren't permanents.